# Introduction to Programming with Python - Part 3

##  Setup

With this Google Colaboratory (Colab) notebook open, click the "Copy to Drive" button that appears in the menu bar. The notebook will then be attached to your own user account, so you can edit it in any way you like -- you can even take notes directly in the notebook.

## Instructors

- Ashley Evans Bandy
- Claire Cahoon
- Walt Gurley
- Natalia Lopez

## Learning objectives

By the end of our workshop today, we hope you'll build on basic Python syntax to understand functions and reading and writing files. With these in hand, you'll know enough to write and apply basic scripts and explore other features of the language.  

## Today's Topics
- Functions
- Reading and writing files

If you would like to learn more about data analysis using Python, we will explore these topics in our Python Open Lab series.

For a reference to previous workshop content, you can access a version of the previous workshop materials with all example and exercise code completed in the [Introduction to Programming with Python - Part 1 filled notebook](https://colab.research.google.com/github/NCSU-Libraries/data-viz-workshops/blob/master/Introduction_to_Programming_with_Python/filled_Introduction_to_Programming_with_Python_1.ipynb).

TODO: update this link

## Questions during the workshop

Please feel free to ask questions throughout the workshop.

We have a second instructor who will available during the workshop. They will answer as able, and will collect questions with answers that might help everyone to be answered at the end of the workshop.

## Functions

At the most basic level, functions are chunks of reusable code. A function can accept data, passed in as arguments, and return data.

To learn more about functions:

- ["Functions" chapter of *A Byte of Python*](https://python.swaroopch.com/functions.html)
- [Python documentation for defining functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)

In [19]:
# Define a function that prints out a hello message -- note use of indenting!
def print_hello():
  print("Hello, it's nice to see you!")

# Call the print_hello function to run the code contained in the function
print_hello()

# Print the type of the value returned from calling the print_hello function
type(print_hello())

Hello, it's nice to see you!
Hello, it's nice to see you!


NoneType

### Returning data from a function

The `hello` function above does not return any data, it only calls a `print` statement. We cannot do anything with this function beyond printing a message. We can use the special `return` keyword to return a value when the function is called

In [20]:
# Define a function that returns a hello message
def return_hello():
    return "Hello, it's nice to see you!"

# Call the return_hello function
return_hello()

# Print the type of the return value of the return_hello function
type(return_hello())

str

Return values can be stored in variables and used just like other values

In [21]:
hello = return_hello()
hello.upper()

"HELLO, IT'S NICE TO SEE YOU!"

### Passing arguments to a function

In [ ]:
# Define a function that returns a hello message
def return_hello_personal(name, workshop):
    return "Hello " + name + ", it's nice to see you at " + workshop + " today!"

### Using functions

In [22]:
# Define a function that replaces periods with exclamation marks in a string
def replace_period(string):
  new_string = ""
  for char in string:
    if char == ".":
      new_string = new_string + "!"
    else:
      new_string = new_string + char
  return new_string

# Call the function on a string that contains periods
replace_period("This. is. a. boring. sentence.")

'This! is! a! boring! sentence!'

**Try it yourself:** Create a function named `add_nums` that takes two numbers as arguments and returns the results of adding these two numbers

In [1]:
# A function that returns the results of adding two numbers passed as arguments
def add_nums(num_1, num_2):
    return num_1 + num_2

add_nums(5, 3)

8

**Bonus Try it yourself:** Create a function named `sum_nums` that takes a list of numbers and returns the sum of those humbers


In [4]:
# A function that returns the sum of a list of numbers
def sum_nums(nums):
    running_sum = 0
    for num in nums:
        running_sum += num
    return running_sum
    
sum_nums([10, 20, 20, 50])

100

## Reading and writing files

Working with comma-separated and similar data files will be covered in a later workshop. It's worthwhile, however, to see how to read and write data or text to and from a file. We'll start with writing some text to a file, then explore how to read it.

### Write to a file

In [5]:
# Sample text to write to a file
sample_text = """
Cats and kittens everywhere,
Hundreds of cats,
Thousands of cats,
Millions and billions and trillions of cats.
"""

# Write the sample text to a file (test different open modes)
with open('cat_text.txt', 'w') as f:
    f.write(sample_text)
    
# You might sometimes see an older pattern:
# f = open('lorem.txt', 'w')
# f.write(sample_text)
# f.close()

You can check the "Files" tab in the column at left now to find the output file.
Note that you must click the REFRESH" button to see it.

### Read from a file

In [6]:
# Read the newly created file.
with open('cat_text.txt', 'r') as f:
    print(f.read())


Cats and kittens everywhere,
Hundreds of cats,
Thousands of cats,
Millions and billions and trillions of cats.



In [7]:
# We can also read the file line by line
with open('cat_text.txt', 'r') as f:
    for line in f:
        line = line.replace('\n', 'END')
        print(line)

END
Cats and kittens everywhere,END
Hundreds of cats,END
Thousands of cats,END
Millions and billions and trillions of cats.END


#### Parsing data read from a file

External data we read into our application is often not formatted for our desired manipulations or analyses. We often have to parse the data from a file into a structure that is appropriate for manipulation or analysis.

In [8]:
# Create an empty list to store the data we will parse
data = []

# Read in the cat_text file line by line
with open('cat_text.txt', 'r') as f:
    for line in f:
        line = line.rstrip('\n')
        line = line.split(' ')
        data.append(line)

# Print out the resulting list
data

[[''],
 ['Cats', 'and', 'kittens', 'everywhere,'],
 ['Hundreds', 'of', 'cats,'],
 ['Thousands', 'of', 'cats,'],
 ['Millions', 'and', 'billions', 'and', 'trillions', 'of', 'cats.']]

### Activity: Parse a text file containing data seperated by semicolons (10 min)

We have membership information for 25 members of the *International Cats of Mystery* service stored in the file `international_cats_of_mystery.txt`. The file is formatted as follows:

1. Each line is seperated by a newline character (`\n`)

1. The first line in the file contains the data variable names (e.g., Name, Joined, Subscriptions)

1. Each subsequent line contains the data variable values for one member (e.g., Bill, 2015, Little Known Hiss-tories podcast)

1. Each variable name or value is seperated by semicolons (`;`)

Use a file parsing method to parse the `international_cats_of_mystery.txt` file to create a list that contains lists of values from each line in the file. The first two items in this list should be:

```python
[
  ['Name', 'Joined', 'Subscriptions'],
  ['Bill', '2015', 'Little Known Hiss-tories podcast'],
  ...
]
```

In [13]:
# Fetch the text file for the activity
!curl https://raw.githubusercontent.com/ncsu-libraries-data-vis/introduction-to-programming-with-python/main/International_cats_of_mystery_US_chapter.txt -o international_cats_of_mystery.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2667  100  2667    0     0  13073      0 --:--:-- --:--:-- --:--:-- 13137


In [24]:
# Parse the data in the "international_cats_of_mystery.txt" file
member_data = []

with open('international_cats_of_mystery.txt') as f:
    for line in f:
        line = line.rstrip('\n')
        member_data.append(line.split(';'))

member_data[:3]

[['Name',
  'Joined',
  'Address 1',
  'Address 2',
  'City',
  'State',
  'Zip Code',
  'Subscriptions'],
 ['Bill',
  '2015',
  '12309 Scratch Tree Lane',
  'Apt. 1',
  'Beverly Hills',
  'CA',
  '90210',
  'Little Known Hiss-tories podcast'],
 ['Pete',
  '2018',
  '501 Grumpy Street',
  '',
  'Long Beach',
  'California',
  '90803',
  'Leisure Cats magazine, Little Known Hiss-tories podcast']]

### Final Activity

Use a loop and a function to create a personalized message for each member of the *International Cats of Mystery* service from the text file parsed in the previous cell. The personalized message should use this template:

> *Hello **[member name]**, Thank you for being a valued member for the last **[member's years of membership]** years. We have an exclusive discount on our **[member's top subscription with the service]**.*

In [25]:
# Create a function that takes in a member's information in the form of the
# parsed data from the "international_cats_of_mystery.txt" file and returns a
# personalized message
def personalized_message(member_info):
    name = member_info[0]
    member_years = 2021 - int(member_info[1])
    top_subscription = member_info[2].split(",")[0]
    return ("Hello " + name
    + ", Thank you for being a valued member for the last " + str(member_years)
    + " years. We have an exclusive discount on our " + top_subscription)

In [26]:
# Loop through the parsed data from the "international_cats_of_mystery.txt" file
# and print a personalized message for each member
for member in member_data[2:]:
    print(personalized_message(member))

Hello Pete, Thank you for being a valued member for the last 3 years. We have an exclusive discount on our 501 Grumpy Street
Hello Bob, Thank you for being a valued member for the last 2 years. We have an exclusive discount on our 62 Ocean Catch Way
Hello Carla, Thank you for being a valued member for the last 15 years. We have an exclusive discount on our 2120 Curious Court
Hello Jose, Thank you for being a valued member for the last 1 years. We have an exclusive discount on our 10036 Sunny Spot Avenue
Hello Ali, Thank you for being a valued member for the last 14 years. We have an exclusive discount on our 143 Right Meow Place
Hello Marcus, Thank you for being a valued member for the last 2 years. We have an exclusive discount on our 21 Bird Watch Circle
Hello Tanvi, Thank you for being a valued member for the last 4 years. We have an exclusive discount on our 4532 Tiny Mouse Street
Hello Lian, Thank you for being a valued member for the last 13 years. We have an exclusive discount o

## Further resources and topics

### Filled version of this notebook

[Introduction to Programming with Python - Part 3 filled notebook](https://colab.research.google.com/github/NCSU-Libraries/data-viz-workshops/blob/master/Introduction_to_Programming_with_Python/Introduction_to_Programming_with_Python_2_filled.ipynb) - a version of this notebook with all code filled in for the guided activity and exercises.
TODO - update this link

### Python Open Labs

Join us next time for the start of the Python Open Labs, which will cover the Pandas data ayalysis library. The format for the Open Labs will include a short lesson demonstrating a topic and extended time for practice activities and discussion.

### Resources

- [A Byte of Python](https://python.swaroopch.com/) is a great intro book and reference for Python
- [Official Python documentation and tutorials](https://docs.python.org/3/)
- [Real Python](https://realpython.com/) contains a lot of different tutorials at different levels
- [LinkedIn Learning](https://www.lynda.com/Python-training-tutorials/415-0.html) is free with NC State accounts and contains several video series for learning Python
- [Dataquest](https://www.dataquest.io/) is a free then paid series of courses with an emphasis on data science

### Topics

- Other data structures: sets, tuples
- Libraries, packages, and pip
- Virtual environments
- Text editors and local execution environments
- The object-oriented paradigm in Python: classes, methods

### Installing Python 

There are quite a few ways to install Python on your own computer, including the [official Python downloads](https://www.python.org/downloads/) and the very popular data-science focused [Anaconda Python distribution](https://www.anaconda.com/products/individual). Depending on your operating system, how you want to write code, and what type of projects you might work on, there are other approaches as well, such as using [pyenv](https://github.com/pyenv/pyenv) and [poetry](https://python-poetry.org/). If you're not sure which approach to take, feel free to get in touch and we'll talk through options and help you get set up. 

### Popular editors for Python

Today we've been writing and running code in Google Colab, which is one particular version of Jupyter Notebooks. Depending on your projects and what you're working on, you may want to write your code in a text editor. While there are many options, if you're just getting started we recommend [Visual Studio Code](https://code.visualstudio.com/) for any operating system but are happy to talk through other editors.


## Further resources and topics

#### Unfilled version of this notebook

[Introduction to Programming with Python - Part 3 unfilled notebook](https://colab.research.google.com/github/ncsu-libraries-data-vis/introduction-to-programming-with-python/blob/main/Introduction_to_Programming_with_Python_3.ipynb) - a version of this notebook without code filled in for the guided activity and exercises. Use the unfilled version to learn these materials or lead a workshop session.

### Resources

- [A Byte of Python](https://python.swaroopch.com/) is a great intro book and reference for Python
- [Official Python documentation and tutorials](https://docs.python.org/3/)
- [Real Python](https://realpython.com/) contains a lot of different tutorials at different levels
- [LinkedIn Learning](https://www.lynda.com/Python-training-tutorials/415-0.html) is free with NC State accounts and contains several video series for learning Python
- [Dataquest](https://www.dataquest.io/) is a free then paid series of courses with an emphasis on data science

### Topics

- Other data structures: sets, tuples
- Libraries, packages, and pip
- Virtual environments
- Text editors and local execution environments
- The object-oriented paradigm in Python: classes, methods

### Installing Python 

There are quite a few ways to install Python on your own computer, including the [official Python downloads](https://www.python.org/downloads/) and the very popular data-science focused [Anaconda Python distribution](https://www.anaconda.com/products/individual). Depending on your operating system, how you want to write code, and what type of projects you might work on, there are other approaches as well, such as using [pyenv](https://github.com/pyenv/pyenv) and [poetry](https://python-poetry.org/). If you're not sure which approach to take, feel free to get in touch and we'll talk through options and help you get set up. 

### Popular editors for Python

Today we've been writing and running code in Google Colab, which is one particular version of Jupyter Notebooks. Depending on your projects and what you're working on, you may want to write your code in a text editor. While there are many options, if you're just getting started we recommend [Visual Studio Code](https://code.visualstudio.com/) for any operating system but are happy to talk through other editors.


## Evaluation survey
Please, spend 1 minute answering these questions that can help us a lot on future workshops. 

[go.ncsu.edu/dvs-eval](https://go.ncsu.edu/dvs-eval)

## Credits

These materials were developed by Claire Cahoon and Walt Gurley from the NC State University Libraries. Materials are based on workshops by Scott Bailey, Ashley Evans Bandy, Natalia Lopez, Vincent Tompkins, Javier de la Rosa, Peter Broadwell, and Simon Wiles.

## Additional Practice Activity:

### Generate a dictionary and personalized messages using data from a file

This activity incorporates concepts we have covered in all *Introduction to Programming with Python* workshops. Use the skills you have developed through your work on these materials as well as references to additional resources we provide to complete this extended activity.

For this activity we have membership information for 25 members of the *International Cats of Mystery* service stored in the file `international_cats_of_mystery.txt`. The file is formatted as follows:

1. Each line is seperated by a newline character (`\n`)

1. The first line in the file contains the data variable names (e.g., Name, Joined, Address 1,...)

1. Each subsequent line contains the data variable values for one member (e.g., Bill, 2015, 12309 Scratch Tree Lane,...)

1. Each variable name or value is seperated by semicolons (`;`)

Our job is to read in the `international_cats_of_mystery.txt` file, parse the data, and format the data to create a list of dictionaries containing membership information that match the format:

```python
{'address 1': '12309 Scratch Tree Lane',
  'address 2': 'Apt. 1',
  'city': 'Beverly Hills',
  'joined': 2015,
  'name': 'Bill',
  'state': 'CA',
  'subscriptions': ['Little Known Hiss-tories podcast'],
  'zip code': '90210'}
```

Once we have the data prepared we will generate the following personalized message for each member:

> *Hello **[member name]**, Thank you for being a valued member of International Cats of Mystery for the last **[member's years of membership]** years. We have an exclusive discount on our **[member's top subscription with the service]**.*

This is similar to the final activity in *Introduction to Programming with Python 1* except this time we will use a `for` loop to automate this process.

In [27]:
# Fetch the text file for the final activity
!curl https://raw.githubusercontent.com/NCSU-Libraries/data-viz-workshops/master/Introduction_to_Programming_with_Python/International_cats_of_mystery_US_chapter.txt -o international_cats_of_mystery.txt

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2667  100  2667    0     0  11697      0 --:--:-- --:--:-- --:--:-- 11646


In [28]:
# Read whole text file to preview contents
with open('international_cats_of_mystery.txt', 'r') as f:
    print(f.read())

Name;Joined;Address 1;Address 2;City;State;Zip Code;Subscriptions
Bill;2015;12309 Scratch Tree Lane;Apt. 1;Beverly Hills;CA;90210;Little Known Hiss-tories podcast
Pete;2018;501 Grumpy Street;;Long Beach;California;90803;Leisure Cats magazine, Little Known Hiss-tories podcast
Bob;2019;62 Ocean Catch Way;;Astoria;OR;97103;email newsletter
Carla;2006;2120 Curious Court;unit B;Raleigh;North Carolina;27610;email newsletter, Cat Nips subscription box
Jose;2020;10036 Sunny Spot Avenue;;Greensboro;NC;27502;Cat Nips subscription box, Little Known Hiss-tories podcast, Leisure Cats magazine
Ali;2007;143 Right Meow Place;Apt. 2;Portland;ME;3901;email newsletter, Little Known Hiss-tories podcast
Marcus;2019;21 Bird Watch Circle;;Austin;Texas;73301;Leisure Cats Magazine, Little Known Hiss-tories podcast
Tanvi;2017;4532 Tiny Mouse Street;Unit D;Dallas;TX;73303;Little Known Hiss-tories podcast, Cat Nips subscription box
Lian;2008;3132 Goldfish Lane;;Ithaca;NY;14850;email newsletter
Arturo;2014;5120 Ki

### Read and parse data line by line and store in a list

Read the membership information data file line by line, parse the string (remove the newline character and split based on semicolon), and store each line as a list in the empty list `data_from_file`.

Tip: Remember the list method `append()` to append new items to a list

In [29]:
# Empty list to store data
data_from_file = []

# Read in text file line by line
with open('international_cats_of_mystery.txt') as f:
    for line in f:
        line = line.rstrip('\n')
        data_from_file.append(line.split(';'))

# Print data_from_file
data_from_file[:3]

[['Name',
  'Joined',
  'Address 1',
  'Address 2',
  'City',
  'State',
  'Zip Code',
  'Subscriptions'],
 ['Bill',
  '2015',
  '12309 Scratch Tree Lane',
  'Apt. 1',
  'Beverly Hills',
  'CA',
  '90210',
  'Little Known Hiss-tories podcast'],
 ['Pete',
  '2018',
  '501 Grumpy Street',
  '',
  'Long Beach',
  'California',
  '90803',
  'Leisure Cats magazine, Little Known Hiss-tories podcast']]

### Create member info dictionaries

We now have a list containing one list per line from the file that we read (stored in `data_from_file`). We need to loop through the `data_from_file` list and use the variable names and values to create dictionaries containing each member's information.

To do this we will loop over the lists containing data in `data_from_file` (`data_from_file[1:]`, remember the first list item containes the variable names), create an empty dictionary, and then use another loop to loop over a range of values the length of items in `data_from_file[0]` to create a dictionary for each member consisting of variable names (`data_from_file[0][i]`) and values (`data[i]`) as key, value pairs.

```python
member_info = []

for data in data_from_file[1:]:
    data_dict = {}
    for i in range(len(data_from_file[0])):
        data_dict[data_from_file[0][i]] = data[i]
    member_info.append(data_dict)
```

In [30]:
# Empty list to store dictionaries
member_info = []

# Loop over each data list from the parsed data
for data in data_from_file[1:]:
    data_dict = {}
    for i in range(len(data_from_file[0])):
        data_dict[data_from_file[0][i]] = data[i]
    member_info.append(data_dict)

# Print member_info
member_info[:3]

[{'Name': 'Bill',
  'Joined': '2015',
  'Address 1': '12309 Scratch Tree Lane',
  'Address 2': 'Apt. 1',
  'City': 'Beverly Hills',
  'State': 'CA',
  'Zip Code': '90210',
  'Subscriptions': 'Little Known Hiss-tories podcast'},
 {'Name': 'Pete',
  'Joined': '2018',
  'Address 1': '501 Grumpy Street',
  'Address 2': '',
  'City': 'Long Beach',
  'State': 'California',
  'Zip Code': '90803',
  'Subscriptions': 'Leisure Cats magazine, Little Known Hiss-tories podcast'},
 {'Name': 'Bob',
  'Joined': '2019',
  'Address 1': '62 Ocean Catch Way',
  'Address 2': '',
  'City': 'Astoria',
  'State': 'OR',
  'Zip Code': '97103',
  'Subscriptions': 'email newsletter'}]

The following cell uses list comprehension to perform the same operations as the previous cell. This is included to show a comparison of two methods for generating the same resultes. The list comprehension uses a `zip` function to create an iterable of paired tuples of variable names (from `data_from_file[0]`) and variable values from each member data row in `data_from_file` (i.e., `data_from_file[1:]`). The `dict()` function is then used to construct a dictionary from the output of `zip()`.

In [31]:
# For list items starting at index 1 in data_lists create a dictionary of
# variable names (data_lists[0]) and values as key, value pairs
member_info = [dict(zip(data_from_file[0], values)) for values in data_from_file[1:]]

member_info[:3]

[{'Name': 'Bill',
  'Joined': '2015',
  'Address 1': '12309 Scratch Tree Lane',
  'Address 2': 'Apt. 1',
  'City': 'Beverly Hills',
  'State': 'CA',
  'Zip Code': '90210',
  'Subscriptions': 'Little Known Hiss-tories podcast'},
 {'Name': 'Pete',
  'Joined': '2018',
  'Address 1': '501 Grumpy Street',
  'Address 2': '',
  'City': 'Long Beach',
  'State': 'California',
  'Zip Code': '90803',
  'Subscriptions': 'Leisure Cats magazine, Little Known Hiss-tories podcast'},
 {'Name': 'Bob',
  'Joined': '2019',
  'Address 1': '62 Ocean Catch Way',
  'Address 2': '',
  'City': 'Astoria',
  'State': 'OR',
  'Zip Code': '97103',
  'Subscriptions': 'email newsletter'}]

### Convert the `Joined` key value from a string to an integer

When we read in our file, all of the data was parsed as a string. We must convert any data that should be in another data type accordingly.

In [32]:
# Test that value stored in 'Joined' is a string
# for member in member_info:
#     print(type(member['Joined']))

# Convert the 'Joined' value to an integer for each member
for member in member_info:
    member['Joined'] = int(member['Joined'])

member_info[:3]

[{'Name': 'Bill',
  'Joined': 2015,
  'Address 1': '12309 Scratch Tree Lane',
  'Address 2': 'Apt. 1',
  'City': 'Beverly Hills',
  'State': 'CA',
  'Zip Code': '90210',
  'Subscriptions': 'Little Known Hiss-tories podcast'},
 {'Name': 'Pete',
  'Joined': 2018,
  'Address 1': '501 Grumpy Street',
  'Address 2': '',
  'City': 'Long Beach',
  'State': 'California',
  'Zip Code': '90803',
  'Subscriptions': 'Leisure Cats magazine, Little Known Hiss-tories podcast'},
 {'Name': 'Bob',
  'Joined': 2019,
  'Address 1': '62 Ocean Catch Way',
  'Address 2': '',
  'City': 'Astoria',
  'State': 'OR',
  'Zip Code': '97103',
  'Subscriptions': 'email newsletter'}]

### Fix state names using a defined function

Some of the state names in our member info list are not formatted correctly as two letter abbreviations. We will create a function to fix the incorrectly formatted state names.

In [33]:
# State abbreviations lookup
state_abbreviations = {
    'California': 'CA',
    'New York': 'NY',
    'North Carolina': 'NC',
    'Texas': 'TX'
}

# Fix state names function
def fix_state_name(member_state):
    if len(member_state) > 2:
        return state_abbreviations[member_state]
    return member_state

In [34]:
# Run the fix_state_format for each member in member_info
for member in member_info:
    member['State'] = fix_state_name(member['State'])

member_info[:3]

[{'Name': 'Bill',
  'Joined': 2015,
  'Address 1': '12309 Scratch Tree Lane',
  'Address 2': 'Apt. 1',
  'City': 'Beverly Hills',
  'State': 'CA',
  'Zip Code': '90210',
  'Subscriptions': 'Little Known Hiss-tories podcast'},
 {'Name': 'Pete',
  'Joined': 2018,
  'Address 1': '501 Grumpy Street',
  'Address 2': '',
  'City': 'Long Beach',
  'State': 'CA',
  'Zip Code': '90803',
  'Subscriptions': 'Leisure Cats magazine, Little Known Hiss-tories podcast'},
 {'Name': 'Bob',
  'Joined': 2019,
  'Address 1': '62 Ocean Catch Way',
  'Address 2': '',
  'City': 'Astoria',
  'State': 'OR',
  'Zip Code': '97103',
  'Subscriptions': 'email newsletter'}]

### Convert `Subscriptions` string to a list

Finally, as our file was parsed as a string, the values contained in the `Subscriptions` key are not formatted as a list. We must convert the current string form to a list using the string function `split()`.

In [35]:
# Convert the Subscriptions key value for each member to a list
for member in member_info:
    member['Subscriptions'] = member['Subscriptions'].split(',')

member_info[:3]

[{'Name': 'Bill',
  'Joined': 2015,
  'Address 1': '12309 Scratch Tree Lane',
  'Address 2': 'Apt. 1',
  'City': 'Beverly Hills',
  'State': 'CA',
  'Zip Code': '90210',
  'Subscriptions': ['Little Known Hiss-tories podcast']},
 {'Name': 'Pete',
  'Joined': 2018,
  'Address 1': '501 Grumpy Street',
  'Address 2': '',
  'City': 'Long Beach',
  'State': 'CA',
  'Zip Code': '90803',
  'Subscriptions': ['Leisure Cats magazine',
   ' Little Known Hiss-tories podcast']},
 {'Name': 'Bob',
  'Joined': 2019,
  'Address 1': '62 Ocean Catch Way',
  'Address 2': '',
  'City': 'Astoria',
  'State': 'OR',
  'Zip Code': '97103',
  'Subscriptions': ['email newsletter']}]

### Use a `for` loop to generate a pesonalized message for each member

Use the code cells below to write two functions that produce a personalized email with the provided membership information and then loop through the `member_info` list to produce this message for all members.

Tips:

- Define a function that calculates how many years someone has been a member - Define a second function that:
  1. gets the member name,
  1. calls the previous function to calculate their years of membership,
  1. gets their most popular subscription,
  1. and formats this information into a personalized message.

- The year a member joined is an integer. Remember, you can only concatenate strings (recall the `str()` method)

- Subscriptions are a list. Each member's subscriptions are listed from most popular to least popular.

In [36]:
# Define a function that calculates membership years
def calculate_member_years(year_joined):
  return 2021 - year_joined

For reference, here is the message format again:
> *Hello **[member name]**, Thank you for being a valued member of International Cats of Mystery for the last **[member's years of membership]** years. We have an exclusive discount on our **[member's top subscription with the service]**.*

In [37]:
# Define a function that creates the personalized message
def create_message(member):
  membership_years = calculate_member_years(member["Joined"])
  return ("Hello "
    + member["Name"]
    + ", Thank you for being a valued member of International Cats of Mystery for the last "
    + str(membership_years)
    + " years. We have an exclusive discount on our "
    + member["Subscriptions"][0])

In [38]:
# Loop through the "member_info" list to produce the personalized message for
# each member
for member in member_info:
    print(create_message(member))

Hello Bill, Thank you for being a valued member of International Cats of Mystery for the last 6 years. We have an exclusive discount on our Little Known Hiss-tories podcast
Hello Pete, Thank you for being a valued member of International Cats of Mystery for the last 3 years. We have an exclusive discount on our Leisure Cats magazine
Hello Bob, Thank you for being a valued member of International Cats of Mystery for the last 2 years. We have an exclusive discount on our email newsletter
Hello Carla, Thank you for being a valued member of International Cats of Mystery for the last 15 years. We have an exclusive discount on our email newsletter
Hello Jose, Thank you for being a valued member of International Cats of Mystery for the last 1 years. We have an exclusive discount on our Cat Nips subscription box
Hello Ali, Thank you for being a valued member of International Cats of Mystery for the last 14 years. We have an exclusive discount on our email newsletter
Hello Marcus, Thank you for